src/python/pants/backend/cargo_bootstrap/subsystems/rustup.py

```diff
@@ -0,0 +1,290 @@
+  def subsystem_dependencies(cls):
+    return super(Rustup, cls).subsystem_dependencies() + (
+      CMakeForGRPC.scoped(cls),
+      GoForGRPC.scoped(cls),
+      ProtocForGRPC.scoped(cls),
+    )
+
+  @classmethod
+  def register_options(cls, register):
+    super(Rustup, cls).register_options(register)
+    register('--toolchain-version', type=str, default='1.30.0',
+             help='Version of the rust toolchain to use.')
+    register('--rust-components', type=list_option, default=[
+      'rustfmt-preview',
+      'rust-src',
+      'clippy-preview',
+    ], help='Components of the rust toolchain to install.')
+    register('--cargo-version', type=str, default='0.2.1',
+             help='Version of the cargo build tool to use.')
+
+  @memoized_property
+  def update_request(self):
+    opts = self.get_options()
+    version = opts.toolchain_version
+    components = opts.rust_components
+    return RustUpdateRequest(
+      toolchain_root=self._toolchain_root,
+      toolchain_version=text_type(version),
+      rust_components=tuple(components),
+    )
+
+  @memoized_property
+  def cargo_version(self):
+    return self.get_options().cargo_version
+
+  @memoized_property
+  def _toolchain_root(self):
+    return os.path.join(get_pants_cachedir(), 'rust')
+
+  @memoized_property
+  def _cargo_home(self):
+    return os.path.join(self._toolchain_root, 'cargo')
+
+  @memoized_property
+  def _rustup_home(self):
+    return os.path.join(self._toolchain_root, 'rustup')
+
+  @memoized_property
+  def _cargo_bin_dir(self):
+    return os.path.join(self._cargo_home, 'bin')
+
+  @memoized_property
+  def pants_owned_rustup_path(self):
+    return os.path.join(self._cargo_bin_dir, 'rustup')
+
+  @memoized_property
+  def pants_owned_rustup_globs(self):
+    return PathGlobsAndRoot(PathGlobs(['rustup']), self._cargo_bin_dir)
+
+  @memoized_property
+  def cargo_ensure_installed_path(self):
+    return os.path.join(self._cargo_bin_dir, 'cargo-ensure-installed')
+
+  @memoized_property
+  def cargo_ensure_installed_globs(self):
+    return PathGlobsAndRoot(PathGlobs(['cargo-ensure-installed']), self._cargo_bin_dir)
+
+  @memoized_property
+  def _cmake(self):
+    return CMakeForGRPC.scoped_instance(self)
+
+  @memoized_property
+  def _go_dist(self):
+    return GoForGRPC.scoped_instance(self)
+
+  @memoized_property
+  def _protoc(self):
+    return ProtocForGRPC.scoped_instance(self)
+
+  @memoized_method
+  def rustup_exec_env(self):
+    return {
+      'CARGO_HOME': self._cargo_home,
+      'RUSTUP_HOME': self._rustup_home,
+      'PATH': create_path_env_var([
+        self._cmake.bin_dir(),
+        self._go_dist.bin_dir(),
+        self._protoc.bin_dir(),
+        # TODO: execute this in a sealed path!
+      ], env=os.environ.copy(), prepend=True),
+    }
+
+
+class RustupExe(datatype([('exe', Snapshot)])): pass
+
+
+@rule(RustupExe, [Select(Rustup)])
+def unpack_rustup(rustup):
+  if not is_executable(rustup.pants_owned_rustup_path):
```

Comment:
@illicitonion : One important missing bit here is that we do not have a local process execution cache, which means that invoking a process to extract a tarball isn't cached, and would run every time.

I think that with that support, a lot more of this would be cacheable in-engine.

Reply:
Opened #6898 about this: it's worth consider whether that is worth doing as a blocker here... and if not, whether we think it would be useful to allow for lowering-latency/increasing parallelism in the `rsc` task.

Reply:
I think a reasonable middle-ground to get this in would be for the `is_executable` checks to exist in the v1 `Task` to decide whether to make a product request for a `RustupExe` from a version, or make a `RustupExe` from a Snapshot/directory path/something.

Then we'd either block migrating from a v1 `Task` to a `@console_rule` on having a local process cache, or acknowledge that this bootstrapping should be rare and not worry about the performance hit of re-extracting a tar file.